### PR TITLE
fix: shipping and taxes grid issue

### PR DESCRIPTION
**Pull request by shantanu-kadam-epi**
Asana link: https://app.asana.com/0/1199380658344177/1200058080958699

image

### CODE REVIEW

File: erpnext/accounts/doctype/shipping_rule/shipping_rule.py [hunk 1]
```diff
@@ -113,7 +113,34 @@ def add_shipping_rule_to_tax_table(self, doc, shipping_amount):
...
 			shipping_charge["tax_amount"] = shipping_amount
 			shipping_charge["description"] = self.label
-			doc.append("taxes", shipping_charge)
+
+			# Check if already tax is present & add new shipping taxes
+			if doc.get("taxes"):
+				taxes_and_charges = []
+				new_tax = doc.get("taxes")
+				for d in doc.get("taxes"):
+					taxes_and_charges.append(d)
```
Comment (hrwX) [inline erpnext/accounts/doctype/shipping_rule/shipping_rule.py]:
```suggestion
				taxes_and_charges = doc.get("taxes")
```

File: erpnext/accounts/doctype/shipping_rule/shipping_rule.py [hunk 2]
```diff
@@ -113,7 +113,34 @@ def add_shipping_rule_to_tax_table(self, doc, shipping_amount):
...
+
+				doc.set("taxes", [])
+				doc.append("taxes", shipping_charge)
+
+				for tax in taxes_and_charges:
+					doc.append("taxes", tax)
```
Comment (hrwX) [inline erpnext/accounts/doctype/shipping_rule/shipping_rule.py]:
```suggestion
				doc.set("taxes", shipping_charge + taxes_and_charges)
```

File: erpnext/accounts/doctype/shipping_rule/shipping_rule.py [hunk 3]
```diff
@@ -113,7 +113,34 @@ def add_shipping_rule_to_tax_table(self, doc, shipping_amount):
...
+			else:
+				doc.append("taxes", shipping_charge)
+
+	# remove existing shipping rule tax by its account head from taxes table
+	def remove_existing_shipping_rule(self, doc, doctype, docname):
+		if not docname:
+			return
+
+		shipping_rule_accounts = frappe.get_all("Shipping Rule", fields=["account"])
+		shipping_rule_names = [data['account'] for data in shipping_rule_accounts]
```
Comment (hrwX) [inline erpnext/accounts/doctype/shipping_rule/shipping_rule.py]:
```suggestion
		shipping_rule_names = [rule.get('account') for rule in frappe.get_all("Shipping Rule", fields=["account"])]
```

File: erpnext/controllers/accounts_controller.py [hunk 4]
```diff
@@ -451,6 +451,7 @@ def clear_unallocated_advances(self, childtype, parentfield):
 	def apply_shipping_rule(self):
 		if self.shipping_rule:
 			shipping_rule = frappe.get_doc("Shipping Rule", self.shipping_rule)
+			shipping_rule.remove_existing_shipping_rule(self,"Shipping Rule", self.shipping_rule)
```
Comment (hrwX) [inline erpnext/controllers/accounts_controller.py]: - why are you giving self argument here ?

File: erpnext/controllers/accounts_controller.py [hunk 5]
```diff
@@ -922,9 +930,21 @@ def get_taxes_and_charges(master_doctype, master_name):
...
+	sales_taxes_accounts = frappe.get_all("Sales Taxes and Charges", fields=["account_head"])
+	sales_taxes_names = [data['account_head'] for data in sales_taxes_accounts]
+
+	if doc_taxes:
+		to_remove = [d for d in doc_taxes if d['account_head'] in sales_taxes_names]
+
+		for d in to_remove:
+			doc_taxes.remove(d)
+
+	return doc_taxes
```
Comment (hrwX) [inline erpnext/controllers/accounts_controller.py]: commonify `remove_existing_sales_taxes` and `remove_existing_shipping_rule` functions.

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/accounts/doctype/shipping_rule/shipping_rule.py (hunk 2 above)
Comment (shantanu-kadam-epi) [reply erpnext/accounts/doctype/shipping_rule/shipping_rule.py]: Shipping charges & taxes charges are in a different format(one is a list and the other is dict) so we cannot append with the + sign, also for loop is used for if taxes & charges are having more than one record.

File: erpnext/controllers/accounts_controller.py (hunk 4 above)
Comment (shantanu-kadam-epi) [reply erpnext/controllers/accounts_controller.py]: Without self, I am not able to access doc to get taxes

File: erpnext/controllers/accounts_controller.py (hunk 5 above)
Comment (shantanu-kadam-epi) [reply erpnext/controllers/accounts_controller.py]: remove_existing_sales_taxes and remove_existing_shipping_rule functions are different.
the remove_existing_sales_taxes returns the taxes after removing the records & taxes get appended on the js side.
remove_existing_shipping_rule doesn't return any taxes & in this function, taxes get appended on the python side.

**Comment by hrwX** [comment]
@shantanu-kadam-epi  I want to test it before merging this PR.

**Comment by shantanu-kadam-epi** [comment]
Created new PR for the same, so closing this PR

**Pull request closed without merging**